### PR TITLE
feat(cli): enable sub-process to run in detached mode

### PR DESCRIPTION
**Pull request by filipporeds**
The PR is to enable processes started via `Deno.run()` to run in detached mode.

This is a continuation of #5836,
Discussion can be found in #5501

TODO:
- [x] Typing checks to ensure that piped is not passed to a detached process
- [x] Check windows output from the CI and figure out how to test it

### CODE REVIEW

File: cli/dts/lib.deno.ns.d.ts [hunk 1]
```diff
@@ -1893,19 +1893,35 @@ declare namespace Deno {
...
     stdout?: "inherit" | "piped" | "null" | number;
     stderr?: "inherit" | "piped" | "null" | number;
     stdin?: "inherit" | "piped" | "null" | number;
   }
 
+  interface RunOptionsDetached extends RunOptionsCommon {
+    detached?: true,
+    stdout?: "null";
+    stderr?: "null";
+    stdin?: "null";
```
Comment (nayeemrmn) [inline cli/dts/lib.deno.ns.d.ts:1922]: Are you sure these options aren't still applicable? What's the runtime behaviour if you pass `"inherit"` vs `"null"` here?

File: cli/tests/integration_tests.rs [hunk 2]
```diff
@@ -3530,3 +3530,54 @@ fn rust_log() {
...
+    .unwrap()
+    .wait_with_output()
+    .unwrap();
+  assert!(output.status.success());
+  assert!(!output.stdout.is_empty());
+
+  let pid = std::str::from_utf8(&output.stdout).unwrap().trim();
+
+  #[cfg(unix)]
+  fn check_runing(pid: &str, running: bool) {
```
Comment (caspervonb) [inline cli/tests/integration_tests.rs:3554]: Minor nit
```suggestion
  fn check_running(pid: &str, running: bool) {
```

File: cli/dts/lib.deno.ns.d.ts (hunk 1 above)
Comment (qu4k) [reply cli/dts/lib.deno.ns.d.ts:1922]: Actually inherit is still good as it's the default if it's undefined. I just forgot to add it as an option.

Running a detached process with inherit will still log the results to the console, even when the process spawner quits.

File: cli/tests/integration_tests.rs (hunk 2 above)
Comment (qu4k) [reply cli/tests/integration_tests.rs:3554]: 🤦‍♂️

File: cli/dts/lib.deno.ns.d.ts (hunk 1 above)
Comment (qu4k) [reply cli/dts/lib.deno.ns.d.ts:1922]: Passing an `"rid"` also works. So I will disallow only `"piped"`

**Comment by qu4k** [comment]
Should I introduce this as an unstable feature?

**Comment by qu4k** [comment]
There are still some issues with windows not properly detaching... it seems like process spawner hangs until child exits even if detached flags are provided.

I based my windows code on this part of the rust std:
https://github.com/rust-lang/rust/blob/6cae28165f0450fd3f100374b26841e458b8cfef/library/std/src/sys/windows/process.rs#L171-L174

**Comment by lucacasonato** [comment]
@piscisaureus Could you take a look? This doesn't seem to work:
https://github.com/denoland/deno/pull/7481/files#diff-451155f201c7ace608f64a3fc506be9aR108-R116

File: cli/ops/process.rs [hunk 3]
```diff
@@ -101,8 +103,28 @@ fn op_run(
...
   }
 
-  // We want to kill child when it's closed
-  c.kill_on_drop(true);
+  #[cfg(windows)]
+  {
+    let mut win_flags: u32 = 0x00000000;
+    if detached {
+      // Detach process on windows
+      win_flags = win_flags | 0x00000008;
```
Comment (piscisaureus) [inline cli/ops/process.rs:111]: * Please use the name of this constant (`winapi::um::winbase::DETACHED_PROCESS`), don't hard code the value.
* If you want `detached` to have the same semantics as it has in node.js, you'll also need to add the `CREATE_NEW_PROCESS_GROUP` flag.

Comment (qu4k) [reply cli/ops/process.rs]: Thanks a lot,

I currently do not have a windows box to test this with so it's a bit difficult for me to judge whether the implementation is working properly or not

**Comment by bartlomieju** [comment]
@qu4k please rebase the PR and let's try to get it landed for `1.6.0`

**Comment by qu4k** [comment]
> @qu4k please rebase the PR and let's try to get it landed for `1.6.0`

I can but this still doesn't work in windows...

**Comment by lirannl** [comment]
Has this been abandoned? I don't really use Windows so I can't help, but I feel like it's pretty important to be able to spawn detached child processes.

**Comment by filipporeds** [comment]
> Has this been abandoned? I don't really use Windows so I can't help, but I feel like it's pretty important to be able to spawn detached child processes.

Well it's not abandoned but i don't have any way to test on windows so i'm blocked

**Comment by bartlomieju** [comment]
@filipporeds thank you for the PR, but I'm going to close it without a merge. Currently it's most likely that we'll adjust subprocess API as described in https://github.com/denoland/deno/discussions/9435 and ability to detach processes will be included in the rewrite. Thanks again!

**Pull request closed without merging**